lib/libp2p_port.ex

```diff
@@ -0,0 +1,140 @@
+defmodule LambdaEthereumConsensus.Libp2pPort do
+  @moduledoc """
+  A GenServer that allows other elixir processes to send and receive commands to/from
+  the LibP2P server in Go. For now, it only supports subscribing and unsubscribing from
+  topics.
+
+  Requests are generated with an ID, which is returned when calling. Those IDs appear
+  in the responses that might be listened to by other processes.
+  """
+
+  use GenServer
+
+  alias Libp2pProto.{Command, Notification, SubscribeToTopic, UnsubscribeFromTopic}
+  require Logger
+
+  @port_name "libp2p_port/libp2p_port"
+
+  ######################
+  ### API
+  ######################
+
+  def start_link([]) do
+    GenServer.start_link(__MODULE__, [], name: __MODULE__)
+  end
+
+  def subscribe_to_topic(topic_name) do
+    id = UUID.uuid4()
+
+    data =
+      Command.encode(%Command{
+        id: id,
+        c: {:subscribe, %SubscribeToTopic{name: topic_name}}
+      })
+
+    GenServer.cast(__MODULE__, {:send, data})
+    {:ok, id}
+  end
+
+  def unsubscribe_from_topic(topic_name) do
+    id = UUID.uuid4()
+
+    data =
+      Command.encode(%Command{
+        id: id,
+        c: {:unsubscribe, %UnsubscribeFromTopic{name: topic_name}}
+      })
+
+    GenServer.cast(__MODULE__, {:send, data})
+    {:ok, id}
+  end
+
+  ########################
+  ### GenServer Callbacks
+  ########################
+
+  @impl GenServer
+  def init([]) do
+    {:ok,
+     %{
+       port: Port.open({:spawn, @port_name}, [:binary]),
+       buffer: <<>>,
+       waiting_for: :size
+     }}
+  end
+
+  @impl GenServer
+  def handle_cast({:send, data}, %{port: port} = state) do
+    send_delimited(port, data)
+    {:noreply, state}
+  end
+
+  @impl GenServer
+  def handle_info({_port, {:data, data}}, %{buffer: buffer, waiting_for: size} = state) do
+    new_state = handle_buffer(buffer <> data, size)
+    {:noreply, state |> Map.merge(new_state)}
+  end
```

Comment:
This could be simplified using the `{:packet, 4}` option in `Port.open`. That way the port driver handles the size header parsing and segmenting of messages (Elixir-side only).

Reply:
It's insane that elixir handles all of this automatically. Simplified the code as you said!